contrib/babelfishpg_common/src/datetimeoffset.c

```diff
@@ -639,6 +640,43 @@ timestamp_datetimeoffset(PG_FUNCTION_ARGS)
 	PG_RETURN_DATETIMEOFFSET(result);
 }
 
+/* timestamptz_datetimeoffset()
+ * Convert timestamp with time zone to datetimeoffset
+ */
+Datum
+timestamptz_datetimeoffset(PG_FUNCTION_ARGS)
+{
+	TimestampTz timestamp = PG_GETARG_TIMESTAMPTZ(0);
+	Timestamp	time;
+	tsql_datetimeoffset *result;
+
+	struct pg_tm tt,
+			   *tm = &tt;
+	fsec_t		fsec;
+	int			tz = 0;
+
+	if (TIMESTAMP_NOT_FINITE(timestamp))
+		time = timestamp;
+	else
+	{
+		if (timestamp2tm(timestamp, &tz, tm, &fsec, NULL, NULL) != 0)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
+					 errmsg("data out of range for datetimeoffset")));
+		if (tm2timestamp(tm, fsec, NULL, &time) != 0)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
+					 errmsg("data out of range for datetimeoffset")));
+	}
+
+	result = (tsql_datetimeoffset *) palloc(DATETIMEOFFSET_LEN);
+	result->tsql_ts = time;
+	result->tsql_tz = (int16) tz / 60;
+	CheckDatetimeoffsetRange(result);
```

Comment:
Can you explain why are we checking for out of range two times here, first in else block and then by CheckDatetimeoffsetRange?

Reply:
**First check**: The check within timestamp2tm() and tm2timestamp() ensures that the timestamp can be correctly converted to and from the pg_tm structure without any issues. If the conversion fails (e.g., due to an invalid date like February 30), an error is thrown at this point.

**Second check**: The CheckDatetime2Range() function ensures that the final timestamp is valid within the specific range constraints of datetimeoffset as defined by the PostgreSQL system. This is a safeguard to ensure that the value meets all the data type's criteria before it is returned.

This double-check ensures that no invalid or out-of-range values are processed or returned, maintaining data integrity. We have done the same for **timestamptz_datetime2** and **timestamptz_datetime** functions as well.

Reply:
We don't have any check like first one in `timestamp_datetimeoffset`. Can you find any testcase where the first check will fail?

1. If yes then we need to fix this in `timestamp_datetimeoffset` as well.
2. and if no then this check is redundant.

Reply:
This is a sanity check required as internally used pg_localtime may produce invalid date due to inconsistencies in time zone. 
Time zone data issues: If there are issues or inconsistencies in the time zone data used by PostgreSQL, it can lead to invalid dates or times being produced by pg_localtime. 
So, this is only required when we are working with timezones and we donot require this in `timestamp_datetimeoffset`. Even PG itself uses this sanity checks, See: timestamp_trunc function of postgresql, So we should not avoid this check.

Reply:
So, you mean PG's `statement_timestamp` function may produce invalid date/time due to inconsistencies in time zone. I don't agree with that, these sanity check are only required when we are modifying the time like `timestamp_trunc` or processing the user input. Here, it seems redundant to me.

Reply:
Wait, these are not checks, we are finding the time and timezone using `timestamp2tm` and `tm2timestamp` function call so it seems good to me. 

The explanation provided earlier by you is completely wrong.

Reply:
Nack, **That was pretty obvious from the function call itself that we are calculating time and timezone** from their, I just specified the need to throw the error when these functions return non-zero result and the scenario where they can return non-zero result, that was what was asked as a question in the first place. And yes, there can be inconsistencies while calculating time from time zone. For example:
`
Daylight Saving Time (DST) transitions: During the transition periods between Daylight Saving Time (DST) and standard time, there can be ambiguous or non-existent times. For example, when transitioning to DST, there is a "missing" hour, and when transitioning back to standard time, there is a "repeated" hour. 
`
In these cases, the pg_localtime function may produce an invalid date or time. So, that's why the error message is also important, we can't just rely on the function call only, which may result in wrong results.

Reply:
`timestamp2tm` and `tm2timestamp` are converters but returns -1 on failure (value out of range), I think Anikait was referring to this as the first check.

Reply:
Correct!!